pkg/kubelet/nodestatus/setters.go

```diff
@@ -481,18 +483,41 @@ func GoRuntime() Setter {
 }
 
 // RuntimeClasses returns a Setter that sets RuntimeClasses on the node.
-func RuntimeClasses(fn func() []kubecontainer.RuntimeHandler) Setter {
+func RuntimeClasses(
+	rcManager *runtimeclass.Manager,
+	handlersFunc func() []kubecontainer.RuntimeHandler,
+) Setter {
 	return func(ctx context.Context, node *v1.Node) error {
-		if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
+		if rcManager == nil || !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
 			return nil
 		}
-		handlers := fn()
-		node.Status.RuntimeClasses = make([]v1.NodeRuntimeClass, len(handlers))
-		for i, h := range handlers {
+		handlers := handlersFunc()
+		handlersMap := make(map[string]kubecontainer.RuntimeHandler, len(handlers))
+		for _, h := range handlers {
+			handlersMap[h.Name] = h
+		}
+
+		classes, err := rcManager.Lister().List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		node.Status.RuntimeClasses = make([]v1.NodeRuntimeClass, len(classes))
+		for i, class := range classes {
+			handlerName, err := rcManager.LookupRuntimeHandler(&class.Name)
+			if err != nil {
+				klog.ErrorS(err, "Failed to look up the runtime handler", "runtimeClassName", class.Name)
+				continue
+			}
+			handler, ok := handlersMap[handlerName]
+			if !ok {
+				klog.ErrorS(nil, "Failed to look up the runtime handler", "runtimeClassName", class.Name,
+					"runtimeHandlerName", handlerName)
+				continue
+			}
 			node.Status.RuntimeClasses[i] = v1.NodeRuntimeClass{
```

Comment:
Wondering if we should have rather returned the list of the RuntimeHandlers, if we are still allowed to change the API

@thockin @haircommander @giuseppe

Reply:
Renamed to RuntimeHandlers